op-proposer/proposer/driver.go

```diff
@@ -269,18 +279,59 @@ func (l *L2OutputSubmitter) FetchL2OOOutput(ctx context.Context) (*eth.OutputRes
 	return output, true, nil
 }
 
-// FetchDGFOutput gets the next output proposal for the DGF.
+// FetchDGFOutput queries the DGF for the latest game and infers whether it is time to make another proposal
+// If necessary, it gets the next output proposal for the DGF, and returns it along with
+// a boolean for whether the proposal should be submitted at all.
 // The passed context is expected to be a lifecycle context. A network timeout
 // context will be derived from it.
-func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
+func (l *L2OutputSubmitter) FetchDGFOutput(ctx context.Context) (*eth.OutputResponse, bool, error) {
+	cCtx, cancel := context.WithTimeout(ctx, l.Cfg.NetworkTimeout)
 	defer cancel()
+	callOpts := &bind.CallOpts{
+		From:    l.Txmgr.From(),
+		Context: cCtx,
+	}
+
+	gameCount, err := l.dgfContract.GameCount(callOpts)
+	if err != nil {
+		l.Log.Warn("Could not query DisputeGameFactory.gameCount()", "err", err)
+		return nil, false, err
+	}
+
+	if gameCount.Cmp(big.NewInt(0)) == 1 {
+		// If there is at least one game, ensure
+		// enough time has lapsed that we need to make another proposal.
+		// If there is not yet any game, a proposal will be made.
+		latestGameIndex := new(big.Int).Sub(gameCount, big.NewInt(1))
+		latestGame, err := l.dgfContract.GameAtIndex(callOpts, latestGameIndex)
+		if err != nil {
+			l.Log.Warn(fmt.Sprintf(
+				"Could not query DisputeGameFactory.GameAtIndex(%d)",
+				latestGameIndex.Int64()), "err", err)
+			return nil, false, err
+		}
+
+		timeSinceLastGame := time.Since(time.Unix(int64(latestGame.Timestamp), 0))
+		if timeSinceLastGame <= l.Cfg.ProposalInterval {
+			l.Log.Info("Duration since last game not past proposal interval", "duration", timeSinceLastGame)
+			return nil, false, nil
+		}
+
+		l.Log.Info("Duration since last game past proposal interval, submitting proposal now", "duration", timeSinceLastGame)
+	}
 
 	blockNum, err := l.FetchCurrentBlockNumber(ctx)
 	if err != nil {
-		return nil, err
+		l.Log.Warn("Could not fetch current block number", "err", err)
+		return nil, false, err
+	}
+	output, err := l.FetchOutput(ctx, blockNum)
+	l.Log.Warn("Could not fetch output at current block number", "err", err, "blockNum", blockNum)
```

Comment:
We should not warn-log an error if we return it. The caller can decide how to handle it (here, the caller also warn-logs). Better to add the context to the error message.

Reply:
[7df2ef0](https://github.com/ethereum-optimism/optimism/pull/11417/commits/7df2ef0da7158efd7bed65a492a2b1a5ef47853f)